In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_excel(r"Telco_customer_churn.xlsx")

In [3]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [6]:
df.shape

(7043, 33)

In [7]:
df.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='object')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 
 16  Internet Service   7043 

In [19]:
#Check missing values
missing_values=df.isnull().sum()
print("Missing values : ",missing_values)

Missing values :  CustomerID                   0
Count                        0
Country                      0
State                        0
City                         0
Zip Code                     0
Lat Long                     0
Latitude                     0
Longitude                    0
Gender                       0
Senior Citizen               0
Partner                      0
Dependents                   0
Tenure Months                0
Phone Service                0
Multiple Lines               0
Internet Service             0
Online Security              0
Online Backup                0
Device Protection            0
Tech Support                 0
Streaming TV                 0
Streaming Movies             0
Contract                     0
Paperless Billing            0
Payment Method               0
Monthly Charges              0
Total Charges                0
Churn Label                  0
Churn Value                  0
Churn Score                  0
CLTV                 

In [21]:
df['Churn Reason'].isnull().sum()

5174

5174 missing values are usually from the Churn Reason column because customers who haven't churned don't have a churn reason.

In [10]:
df.describe()

,Count,Zip Code,Latitude,Longitude,Tenure Months,Monthly Charges,Churn Value,Churn Score,CLTV
count,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,1.0,93521.964646,36.282441,-119.798880,32.371149,64.761692,0.265370,58.699418,4400.295755
std,0.0,1865.794555,2.455723,2.157889,24.559481,30.090047,0.441561,21.525131,1183.057152
min,1.0,90001.000000,32.555828,-124.301372,0.000000,18.250000,0.000000,5.000000,2003.000000
25%,1.0,92102.000000,34.030915,-121.815412,9.000000,35.500000,0.000000,40.000000,3469.000000
50%,1.0,93552.000000,36.391777,-119.730885,29.000000,70.350000,0.000000,61.000000,4527.000000
75%,1.0,95351.000000,38.224869,-118.043237,55.000000,89.850000,1.000000,75.000000,5380.500000
max,1.0,96161.000000,41.962127,-114.192901,72.000000,118.750000,1.000000,100.000000,6500.000000


In [11]:
df.describe(include='object')

,CustomerID,Country,State,City,Lat Long,Gender,Senior Citizen,Partner,Dependents,Phone Service,...,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Total Charges,Churn Label,Churn Reason
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043.0,7043,1869
unique,7043,1,1,1129,1652,2,2,2,2,2,...,3,3,3,3,3,2,4,6531.0,2,20
top,3668-QPYBK,United States,California,Los Angeles,"33.964131, -118.272783",Male,No,No,No,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No,Attitude of support person
freq,1,7043,7043,305,5,3555,5901,3641,5416,6361,...,3095,3473,2810,2785,3875,4171,2365,11.0,5174,192


In [22]:
Duplicate_IDs=df['CustomerID'].duplicated().sum()
print("Duplicate customer id :",Duplicate_IDs)

Duplicate customer id : 0


In [23]:
#Check invalid monhtly charges
invalid_monthly_charge=df[df['Monthly Charges'] < 0]
print("Invalid monthly charge",len(invalid_monthly_charge))

Invalid monthly charge 0


In [24]:
#Check Invalid tenure months
Invalid_tenure_months=df[df['Tenure Months'] < 0]
print("Invalid tenure months ",len(Invalid_tenure_months))

Invalid tenure months  0


In [26]:
#Check if blank data in total charges column
blank_charges = df[df['Total Charges'].isna()]
print("Rows with blank Total Charges:",blank_charges)


Rows with blank Total Charges: Empty DataFrame
Columns: [CustomerID, Count, Country, State, City, Zip Code, Lat Long, Latitude, Longitude, Gender, Senior Citizen, Partner, Dependents, Tenure Months, Phone Service, Multiple Lines, Internet Service, Online Security, Online Backup, Device Protection, Tech Support, Streaming TV, Streaming Movies, Contract, Paperless Billing, Payment Method, Monthly Charges, Total Charges, Churn Label, Churn Value, Churn Score, CLTV, Churn Reason, Service Adoption Score]
Index: []

[0 rows x 34 columns]


In [12]:
#count of customers city wise
df['City'].value_counts()

City
Los Angeles       305
San Diego         150
San Jose          112
Sacramento        108
San Francisco     104
                 ... 
Healdsburg          4
Jenner              4
Philo               4
Point Arena         4
Olympic Valley      4
Name: count, Length: 1129, dtype: int64

In [13]:
df['Payment Method'].value_counts()

Payment Method
Electronic check             2365
Mailed check                 1612
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Name: count, dtype: int64

In [14]:
df['Internet Service'].value_counts()

Internet Service
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

In [15]:
df['Streaming Movies'].value_counts()

Streaming Movies
No                     2785
Yes                    2732
No internet service    1526
Name: count, dtype: int64

In [16]:
df['Contract'].value_counts()

Contract
Month-to-month    3875
Two year          1695
One year          1473
Name: count, dtype: int64

In [17]:
churned_customers=df[df['Churn Value']==1]
churned_customers.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1869 entries, 0 to 1868
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         1869 non-null   object 
 1   Count              1869 non-null   int64  
 2   Country            1869 non-null   object 
 3   State              1869 non-null   object 
 4   City               1869 non-null   object 
 5   Zip Code           1869 non-null   int64  
 6   Lat Long           1869 non-null   object 
 7   Latitude           1869 non-null   float64
 8   Longitude          1869 non-null   float64
 9   Gender             1869 non-null   object 
 10  Senior Citizen     1869 non-null   object 
 11  Partner            1869 non-null   object 
 12  Dependents         1869 non-null   object 
 13  Tenure Months      1869 non-null   int64  
 14  Phone Service      1869 non-null   object 
 15  Multiple Lines     1869 non-null   object 
 16  Internet Service   1869 non-n

Data Engineering
1. BAsed on Tenure value categorize Customers as New Customer, RegularCustomer , Loyal Customer

In [27]:
bins=[-1,12,36,float('inf')]
labels=[ "New Customer", "Regular Customer", "Loyal Customer"]
df["Customer Segment"]=pd.cut(df['Tenure Months'],bins,labels=labels)

In [28]:
df['Customer Segment'].value_counts()

Customer Segment
Loyal Customer      3001
New Customer        2186
Regular Customer    1856
Name: count, dtype: int64

Revenue Segment
Low Value
Medium Value
High Value
Premium
based on Monthly Charges.

In [29]:
# Define thresholds for Monthly Charges (e.g., $0-35 Low, $35-70 Medium, $70-100 High, $100+ Premium)
bins = [18,40,65,90, float('inf')]
labels = ["Low", "Medium", "High", "Premium"]

# Segment customers based on Monthly Charges
df["Revenue Segment"] = pd.cut(df['Monthly Charges'], bins=bins, labels=labels)


In [30]:
df['Revenue Segment'].value_counts()

Revenue Segment
High       2160
Low        1838
Premium    1739
Medium     1306
Name: count, dtype: int64

In [31]:
#Service Adoption Score
services = [
    "Online Security",
    "Online Backup",
    "Device Protection",
    "Tech Support",
    "Streaming TV",
    "Streaming Movies"
]

df["Service Adoption Score"] = (
    (df[services] == "Yes")
    .sum(axis=1)
)

df["Service Adoption Score"].value_counts()

Service Adoption Score
0    2219
3    1118
2    1033
1     966
4     852
5     571
6     284
Name: count, dtype: int64

2219 customers don't use any of the six additional services.
284 customers use all six services.

In [32]:
#Churn Score(highest are the high risk customers)
df.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason', 'Service Adoption Score',
       'Customer Segment', 'Revenue Segment'],
      dtype='object')

In [33]:
df['Churn Score'].describe()

count    7043.000000
mean       58.699418
std        21.525131
min         5.000000
25%        40.000000
50%        61.000000
75%        75.000000
max       100.000000
Name: Churn Score, dtype: float64

In [34]:
bins=[4,30,60,80,100]
labels=['Low Risk','Mediun Risk','High risk','Critical Risk']
df['Risk Category']=pd.cut(df['Churn Score'],bins=bins,labels=labels,include_lowest=True)

In [35]:
df['Risk Category'].value_counts()

Risk Category
Mediun Risk      2523
High risk        2517
Critical Risk    1050
Low Risk          953
Name: count, dtype: int64

In [4]:
df.to_csv("Cleaned_Transformed_Customers.csv")